### PR TITLE
ARROW-17817: [C++] Let ORC compile on MSVC if it is activated

### PR DESCRIPTION
**Pull request by LouisClt**

### CODE REVIEW

**Comment by kou** [comment]
Could you enable ORC in our CI jobs?

```diff
diff --git a/.github/workflows/cpp.yml b/.github/workflows/cpp.yml
index 2642a6ec1a..07243ffc4e 100644
--- a/.github/workflows/cpp.yml
+++ b/.github/workflows/cpp.yml
@@ -213,6 +213,7 @@ jobs:
       ARROW_HOME: /usr
       ARROW_JEMALLOC: OFF
       ARROW_MIMALLOC: ON
+      ARROW_ORC: ON
       ARROW_PARQUET: ON
       ARROW_USE_GLOG: OFF
       ARROW_VERBOSE_THIRDPARTY_BUILD: OFF
diff --git a/ci/appveyor-cpp-build.bat b/ci/appveyor-cpp-build.bat
index 4b90e49257..d8eb9d2c11 100644
--- a/ci/appveyor-cpp-build.bat
+++ b/ci/appveyor-cpp-build.bat
@@ -68,6 +68,7 @@ cmake -G "%GENERATOR%" %CMAKE_ARGS% ^
       -DARROW_FLIGHT_SQL=%ARROW_BUILD_FLIGHT_SQL% ^
       -DARROW_GANDIVA=%ARROW_BUILD_GANDIVA% ^
       -DARROW_MIMALLOC=ON ^
+      -DARROW_ORC=ON ^
       -DARROW_PARQUET=ON ^
       -DARROW_PYTHON=ON ^
       -DARROW_S3=%ARROW_S3% ^
```

**Comment by LouisClt** [comment]
Hello, I made the changes you suggested. I don't know if this will compile fine as my only compile tests were with vcpkg.

File: ci/appveyor-cpp-build.bat [hunk 1]
```diff
@@ -68,6 +68,7 @@ cmake -G "%GENERATOR%" %CMAKE_ARGS% ^
       -DARROW_FLIGHT_SQL=%ARROW_BUILD_FLIGHT_SQL% ^
       -DARROW_GANDIVA=%ARROW_BUILD_GANDIVA% ^
       -DARROW_MIMALLOC=ON ^
+      -DARROW_ORC=ON
```
Comment (kou) [inline ci/appveyor-cpp-build.bat:71]:
```suggestion
      -DARROW_ORC=ON ^
```

**Review by kou — COMMENTED** [review]
Thanks.

It seems that there are some warnings with Visual C++:

https://github.com/apache/arrow/actions/runs/3111153495/jobs/5043117996#step:11:1325

```text
 D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(49): error C2220: the following warning is treated as an error
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(49): warning C4244: 'initializing': conversion from 'double' to 'int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(49): warning C4244: 'initializing': conversion from 'double' to 'const int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(50): warning C4244: 'initializing': conversion from 'double' to 'int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(50): warning C4244: 'initializing': conversion from 'double' to 'const int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(51): warning C4244: 'initializing': conversion from 'double' to 'int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(51): warning C4244: 'initializing': conversion from 'double' to 'const int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(52): warning C4244: 'initializing': conversion from 'double' to 'int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(52): warning C4244: 'initializing': conversion from 'double' to 'const int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(53): warning C4244: 'initializing': conversion from 'double' to 'int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(53): warning C4244: 'initializing': conversion from 'double' to 'const int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(54): warning C4244: 'initializing': conversion from 'double' to 'int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(54): warning C4244: 'initializing': conversion from 'double' to 'const int64_t', possible loss of data
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(60): error C2061: syntax error: identifier 'ssize_t'
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(61): error C2065: 'capacity': undeclared identifier
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(89): error C2131: expression did not evaluate to a constant
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(89): note: failure was caused by a read of a variable outside its lifetime
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(89): note: see usage of 'length'
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(92): error C3863: array type 'int32_t [length]' is not assignable
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(292): error C2664: 'arrow::MemoryOutputStream::MemoryOutputStream(const arrow::MemoryOutputStream &)': cannot convert argument 1 from 'const int' to 'const arrow::MemoryOutputStream &'
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(292): note: Reason: cannot convert from 'const int' to 'const arrow::MemoryOutputStream'
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(292): note: No constructor could take the source type, or constructor overload resolution was ambiguous
D:\a\arrow\arrow\cpp\src\arrow\adapters\orc\adapter_test.cc(84): note: see declaration of 'arrow::MemoryOutputStream::MemoryOutputStream'
```

Could you fix them? Or do you want me to fix them?

**Comment by LouisClt** [comment]
I can try to fix them, at least for the easy ones. How's that these warnings did not fire before ? It is specific to Visual C++ ?

**Comment by kou** [comment]
I think so.

**Comment by kou** [comment]
Could you apply this for AppVeyor job?

```diff
```

**Comment by LouisClt** [comment]
I made the changes for the AppVeyor job failure. However this still does not work because it can't find a correct version of protobuf (from looking at the logs).
There is also a problem in the "C++ / AMD64 Windows 2019 C++17" job, it seems it can't find some Zstd symbols during linking.

**Comment by kou** [comment]
I've fixed CI failures on AppVeyor and GitHub Actions.

I'll merge this tomorrow if nobody objects it.

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Review by kou — APPROVED** [review]
+1

@davisusanibar FYI: This PR also enables ORC for Windows' JNI.

**Pull request merged**